proposals/p3833.md

```diff
@@ -0,0 +1,157 @@
+
+## Table of contents
+
+-   [Abstract](#abstract)
+-   [Problem](#problem)
+-   [Background](#background)
+-   [Proposal](#proposal)
+    -   [Work to incorporate needed simplifications directly into the design](#work-to-incorporate-needed-simplifications-directly-into-the-design)
+    -   [Keep a full fidelity mode for any other optimizations](#keep-a-full-fidelity-mode-for-any-other-optimizations)
+-   [Rationale](#rationale)
+-   [Alternatives considered](#alternatives-considered)
+    -   [Strictly adhere to a full-fidelity model](#strictly-adhere-to-a-full-fidelity-model)
+    -   [Directly implement the optimized model](#directly-implement-the-optimized-model)
+
+<!-- tocstop -->
+
+## Abstract
+
+The toolchain's [Semantic IR][semir] should start off modeling the full,
+complex, and rich library-based and generic extension point semantics of Carbon
+without eliding any layers or rewrites for compile time efficiency. We shouldn't
+front-load elision or optimization when implementing the designs.
+
+Once we have a full-fidelity implementation, we should work to build an
+efficient elision, short-circuit, or common case simplification into the design
+itself sufficient to make the SemIR model efficient. Only if we cannot find a
+reasonable approach for that should we diverge the SemIR model to optimize its
+efficiency, and we should preserve full fidelity in an optional mode.
+
+[semir]:
+    https://docs.google.com/document/d/1RRYMm42osyqhI2LyjrjockYCutQ5dOf8Abu50kTrkX0/edit?resourcekey=0-kHyqOESbOHmzZphUbtLrTw#heading=h.503m6lfcnmui
+
+## Problem
+
+Carbon's design heavily leverages types and APIs defined in the `Core` package
+and implicitly imported with the prelude. It also defines rich extension points
+and generic semantic models for the language, typically through rewrites from
+the user-facing syntax into more complex syntax that operates through
+`interface`s and `impl`s that provide the generic model and customization
+points.
+
+Modeling this at full fidelity in the toolchain's [Semantic IR][semir] is
+expected to create an unreasonably expensive representation for common and
+pervasive patterns in code. On the other hand, every divergence between the
+SemIR model and the design adds both risk and cost. When we can keep the two in
+sync, we get the best engineering tradeoffs, provided it is reasonably
+efficient.
+
+## Background
+
+Several proposals have started to surface these kinds of challenge, but to a
+lesser degree. Some examples:
+
+-   [Proposal #820 - Implicit conversions](https://github.com/carbon-language/carbon-lang/pull/820)
+-   [Proposal #845 - `as` expressions](https://github.com/carbon-language/carbon-lang/pull/845)
+-   [Proposal #1083 - Arithmetic expressions](https://github.com/carbon-language/carbon-lang/pull/1083)
+
+The most significant step that motivates capturing an explicit strategy here is
+[proposal #3720 - Binding operators](https://github.com/carbon-language/carbon-lang/pull/3720).
+
+## Proposal
```

Comment:
I feel like an example would make this proposal a lot clearer.

Reply:
Good idea, added one below.